scanner/src/main/kotlin/storages/CompositeStorage.kt

```diff
@@ -0,0 +1,115 @@
+ * chain of places where to look up results can be configured.
+ *
+ * Analogously, when a result is to be written the class passes the result to all the storages configured as writers.
+ * For many use cases, a single writer is probably sufficient; but if there are special requirements - e.g. to
+ * create backups of results automatically or to have a fast, but transient cache of results -, it is possible to
+ * specify multiple writers.
+ */
+class CompositeStorage(
+    val readers: List<ScanResultsStorage>,
+    val writers: List<ScanResultsStorage>
+) : ScanResultsStorage() {
+    override val name: String
+        get() {
+            val readerNames = readers.map { it.name }
+            val writerNames = writers.map { it.name }
+            return readerNames.joinToString(separator = ",", prefix = "composite[") +
+                    writerNames.joinToString(separator = ",", prefix = "/", postfix = "]")
+        }
+
+    /**
+     * Try to find scan results for the provided [id]. This implementation iterates over all the reader storages
+     * provided until it receives a non-empty success result.
+     */
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        fetchReadResult(id) { read(id) }
+
+    /**
+     * Try to find scan results for the provided [pkg] and [scannerDetails]. This implementation iterates over all
+     * the reader storages provided until it receives a non-empty success result.
+     */
+    override fun readFromStorage(pkg: Package, scannerDetails: ScannerDetails): Result<ScanResultContainer> =
+        fetchReadResult(pkg.id) { read(pkg, scannerDetails) }
+
+    /**
+     * Trigger all configured writer storages to add the [scanResult] for the given [id]. Return a success result
+     * if at least one writer is successful; otherwise return a failure result with an accumulated error message.
+     */
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> {
+        val writeResults = writers.map { it.add(id, scanResult) }
+        val errors = writeResults.filterIsInstance<Failure<Unit>>()
+        return if (errors.size < writers.size || errors.isEmpty()) Success(Unit)
+        else Failure(errors.joinToString { it.error })
+    }
+
+    /**
+     * Iterate over the configured readers to find a scan result for the given [id] invoking the [readFunc] on each
+     * reader until a result is found. If none of the configured readers returns a defined result, but at least one
+     * returns a success result, return an empty success result. If all readers fail, return a failure result with
+     * an accumulated failure message.
+     */
+    private fun fetchReadResult(id: Identifier, readFunc: ScanResultsStorage.() -> Result<ScanResultContainer>):
+            Result<ScanResultContainer> {
+        if (readers.isEmpty()) return emptyResult(id)
```

Comment:
Am I right that this line is not needed?
If it's just an optimization I believe we should ditch it as it does not matter since we're dealing with I/O which is slow.

Reply:
It is not only an optimization, but an additional condition is needed for returning the correct result. If the line is dropped, the condition if (failures.size < readers.size) in the *return* statement is *false*, and thus a failure result is returned.

Reply:
I see.